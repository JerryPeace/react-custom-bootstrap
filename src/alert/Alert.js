import PropTypes from 'prop-types';
import React from 'react';
import classNames from 'classnames';
import bootstrapUtils from '../utils/bootstrapUtils';
import CloseIcon from '../icon/CloseIcon';
import _ from 'lodash';

const ALERT_TYPE = {
    NORMAL: 'normal',
    TOAST: 'toast'
};

class Alert extends React.Component {

    static propTypes = {
        bsStyle: PropTypes.oneOf(['success', 'info', 'warning', 'danger']),
        onClose: PropTypes.func,
        closeBtn: PropTypes.bool,
        titleText: PropTypes.string,
        bsClass: PropTypes.string,
        isShortMsg: PropTypes.bool,
        closable: PropTypes.bool,
        visible: PropTypes.bool,
        type: PropTypes.oneOf([ALERT_TYPE.NORMAL, ALERT_TYPE.TOAST])
    };

    static defaultProps = {
        bsClass: 'alert',
        bsStyle: 'warning',
        closable: true,
        isShortMsg: false
    };

    constructor (props) {
        super(props);
        this.onClose = this.onClose.bind(this);
        this.state = {
            alertVisible: true
        };
    }

    renderTitle() {
        let inlineStyle = {
            marginRight: '5px'
        };

        const titleText = this.props.titleText;

        return (
            <strong
                className={'tm-alert-title'}
                style={inlineStyle}>
                {titleText}
            </strong>
        );
    }

    renderDismissButton() {
        return (
            <CloseIcon onClick={this.onClose} />
        );
    }

    onClose(e) {
        const { onClose } = this.props;
        const onCloseResult = onClose && onClose(e);

        if (typeof onCloseResult === 'object' && _.isFunction(onCloseResult.then)) {
            onClose(e).then(() => {
                this.close();
            });
        } else {
            this.close();
        }

    }

    close() {
        // it will be controllable if visible has value
        typeof this.props.visible === 'undefined' && this.setState({ alertVisible: false });
    }

    open() {
        this.setState({
            alertVisible: true
        });
    }

    getVisible() {
        const { visible } = this.props;
        if (typeof visible === 'undefined') {
            return this.state.alertVisible;
        } else {
            return visible;
        }
    }

    render() {
        let isClosable = this.props.closable;
        // add specific classes
        let classes = {
            'tm-alert-dismissable': isClosable,
            'tm-alert-block': !this.props.isShortMsg,
            'toast-alert': this.props.type === ALERT_TYPE.TOAST
        };

        let bsCls = bootstrapUtils.getCls(this.props);

        const {titleText, bsClass, bsStyle, closable, isShortMsg, onClose, ...rest} = this.props;

        if (this.getVisible()) {
            return (
                <div {...rest} className={classNames(this.props.className, classes, bsCls)}>
                    {isClosable ? this.renderDismissButton() : null}
                    {titleText ? this.renderTitle() : null}
                    {this.props.children}
                </div>
            );
        }

        return null;
    }
}

Alert.ALERT_TYPE = ALERT_TYPE;

export default Alert;
