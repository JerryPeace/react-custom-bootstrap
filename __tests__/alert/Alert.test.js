import React from 'react';
import { mount } from 'enzyme';
import toJSON from'enzyme-to-json';
import Alert from 'alert/Alert';

describe('Alert', () => {
    let alert;

    describe('with default settings', () => {
        beforeEach(() => {
            alert = mount(<Alert titleText='title-text'>alert-message</Alert>);
        });

        test('should render corrently', () => {
            expect(toJSON(alert)).toMatchSnapshot();
        });

        test('should close it while clicking close button', () => {
            // find close icon
            const closeIcon = alert.find('CloseIcon');
            expect(closeIcon.exists()).toBeTruthy();
            // click it
            closeIcon.simulate('click');
            alert.update();
            // it should hide the alert
            expect(alert.state('alertVisible')).toBeFalsy();
            expect(toJSON(alert)).toMatchSnapshot();
        });
    });

    describe('with onClose callback', () => {
        let onClose;

        beforeEach(() => {
            onClose = jest.fn();
            alert = mount(<Alert titleText='title-text-with-close' onClose={onClose}>
                alert-message</Alert>);
        });

        test('should render corrently', () => {
            expect(toJSON(alert)).toMatchSnapshot();
        });

        test('should call onClose callback while clicking on it', () => {
             // find close icon
            const closeIcon = alert.find('CloseIcon');
            expect(closeIcon.exists()).toBeTruthy();
            // click it
            closeIcon.simulate('click');
            alert.update();
            expect(alert.state('alertVisible')).toBeFalsy();
            expect(toJSON(alert)).toMatchSnapshot();
            expect(onClose.mock.calls.length).toBe(1);
        });
    });

    describe('with in different styles', () => {
        const styles = ['success', 'info', 'warning', 'danger'];

        styles.forEach((style) => {
            test('should render the style with ' + style, () => {
                alert = mount(<Alert titleText='title-text-with-style' bsStyle={style}>
                    alert-message</Alert>);
                expect(alert.find('.alert-' + style).exists()).toBeTruthy();
            });
        });
    });

    test('toast alert should contains toast-alert class name', () => {
        alert = mount(<Alert type='toast' titleText='title-text'>alert-message</Alert>);
        expect(toJSON(alert)).toMatchSnapshot();
    });
});
