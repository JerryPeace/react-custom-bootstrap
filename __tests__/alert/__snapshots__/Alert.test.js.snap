// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert toast alert should contains toast-alert class name 1`] = `
<Alert
  bsClass="alert"
  bsStyle="warning"
  closable={true}
  isShortMsg={false}
  titleText="title-text"
  type="toast"
>
  <div
    className="tm-alert-dismissable tm-alert-block toast-alert alert alert-warning"
    type="toast"
  >
    <CloseIcon
      onClick={[Function]}
    >
      <a
        className="close"
        href="#"
        onClick={[Function]}
      >
        <span>
          ×
        </span>
      </a>
    </CloseIcon>
    <strong
      className="tm-alert-title"
      style={
        Object {
          "marginRight": "5px",
        }
      }
    >
      title-text
    </strong>
    alert-message
  </div>
</Alert>
`;

exports[`Alert with default settings should close it while clicking close button 1`] = `
<Alert
  bsClass="alert"
  bsStyle="warning"
  closable={true}
  isShortMsg={false}
  titleText="title-text"
/>
`;

exports[`Alert with default settings should render corrently 1`] = `
<Alert
  bsClass="alert"
  bsStyle="warning"
  closable={true}
  isShortMsg={false}
  titleText="title-text"
>
  <div
    className="tm-alert-dismissable tm-alert-block alert alert-warning"
  >
    <CloseIcon
      onClick={[Function]}
    >
      <a
        className="close"
        href="#"
        onClick={[Function]}
      >
        <span>
          ×
        </span>
      </a>
    </CloseIcon>
    <strong
      className="tm-alert-title"
      style={
        Object {
          "marginRight": "5px",
        }
      }
    >
      title-text
    </strong>
    alert-message
  </div>
</Alert>
`;

exports[`Alert with onClose callback should call onClose callback while clicking on it 1`] = `
<Alert
  bsClass="alert"
  bsStyle="warning"
  closable={true}
  isShortMsg={false}
  onClose={[Function]}
  titleText="title-text-with-close"
/>
`;

exports[`Alert with onClose callback should render corrently 1`] = `
<Alert
  bsClass="alert"
  bsStyle="warning"
  closable={true}
  isShortMsg={false}
  onClose={[Function]}
  titleText="title-text-with-close"
>
  <div
    className="tm-alert-dismissable tm-alert-block alert alert-warning"
  >
    <CloseIcon
      onClick={[Function]}
    >
      <a
        className="close"
        href="#"
        onClick={[Function]}
      >
        <span>
          ×
        </span>
      </a>
    </CloseIcon>
    <strong
      className="tm-alert-title"
      style={
        Object {
          "marginRight": "5px",
        }
      }
    >
      title-text-with-close
    </strong>
    alert-message
  </div>
</Alert>
`;
